test/NuGet.Clients.Tests/NuGet.PackageManagement.VisualStudio.Test/NetCorePackageReferenceProjectTests.cs

```diff
@@ -447,6 +456,1909 @@ public async Task GetInstalledVersion_WithAssetsFile_ChangingPackageSpec_Returns
+
+                        if (prevPackageSpec != null)
+                        {
+                            packageSpec = packageSpec.WithTestProjectReference(prevPackageSpec);
+                        }
+
+                        // Restore info
+                        var projectRestoreInfo = ProjectJsonTestHelpers.GetDGSpecFromPackageSpecs(packageSpec);
+                        projectCache.AddProjectRestoreInfo(projectNames, projectRestoreInfo, new List<IAssetsLogMessage>());
+                        projectCache.AddProject(projectNames, projectAdapter, project).Should().BeTrue();
+                        prevProj = projectFullPath;
+                        packageSpecs[i] = packageSpec;
+                        prevPackageSpec = packageSpec;
+                        projectFullPaths[i] = projectFullPath;
+                    }
+
+                    var initialInstalledPackages = await netCorePackageReferenceProjects[numberOfProjects-1].GetInstalledPackagesAsync(CancellationToken.None);
+
+                    // Act
+                    var results = await nuGetPackageManager.PreviewProjectsInstallPackageAsync(
+                        netCorePackageReferenceProjects, // All projects
+                        packageB,
+                        resolutionContext,
+                        testNuGetProjectContext,
+                        sourceRepositoryProvider.GetRepositories(),
+                        CancellationToken.None);
+
+                    var actions = results.Select(a => a.Action).ToArray();
+
+                    await nuGetPackageManager.ExecuteNuGetProjectActionsAsync(
+                        netCorePackageReferenceProjects,
+                        actions,
+                        testNuGetProjectContext,
+                        new SourceCacheContext(),
+                        CancellationToken.None);
+
+                    // Assert
+                    Assert.True(telemetryEvents.Count > 1);
+                    var actionTelemetryStepEvents = telemetryEvents.OfType<ActionTelemetryStepEvent>();
+                    Assert.True(actionTelemetryStepEvents.Any(t => t.SubStepName.Contains("Preview build integrated action time")));
+                    var builtIntegratedActions = actions.OfType<BuildIntegratedProjectAction>().ToList();
+                    Assert.Equal(actions.Count(), netCorePackageReferenceProjects .Count());
+                    Assert.Equal(actions.Count(), builtIntegratedActions.Count());
+                    Assert.True(builtIntegratedActions.All(b => b.RestoreResult.Success));
+                }
+            }
+            finally
+            {
+                foreach (TestDirectory projectDirectory in projectDirectories)
+                {
+                    projectDirectory.Dispose();
+                }
+            }
+        }
+
+        [Fact]
+        public async Task TestPacMan_PreviewBuildIntegratedProjectsActionsAsync_UninstallPackageFor_TopAndMidParentProject_Success()
```

Comment:
I know you added a test here, but you are not testing the code you added correct? 

You are calling uninstall sequentially.

Reply:
Here I am only testing correctness part, restore a once per project.
For improvement I created another follow up issue [#9956](https://github.com/NuGet/Home/issues/9956) which will make `create resolved actions` parallel. Now execution of `resolved action` is parallel.